### PR TITLE
Add RQ and QL methods and tests

### PR DESCRIPTION
**Pull request by kshyatt**
Added `orgrq`, `orgql`, `ormrq`, `ormql` and tests for the same. If we are going to have `gerqf` and `geqlf` methods to make the factorizations it's really annoying to not be able to **do** anything with them.

### CODE REVIEW

**Comment by andreasnoack** [comment]
LGTM

File: base/linalg/lapack.jl [hunk 1]
```diff
@@ -1780,6 +1780,74 @@ for (orglq, orgqr, ormlq, ormqr, gemqrt, elty) in
...
+                      &max(1,stride(A,2)), tau, work, &lwork,
+                      info)
+                @lapackerror
+                if lwork < 0
+                    lwork = BlasInt(real(work[1]))
+                    work = Array($elty, lwork)
+                end
+            end
+            if n<size(A,2)
+                A[:,1:n]
```
Comment (jakebolewski) [inline base/linalg/lapack.jl]: I might be good to do an explicit copy here, as it will only have to be changed back for the array views possible change.

Comment (tkelman) [inline base/linalg/lapack.jl]: wouldn't we want the in-place functions to return views wherever possible?

Comment (andreasnoack) [inline base/linalg/lapack.jl]: I think type stability is more important than avoiding the copy. Another option is to always return `A` even though the last columns are garbage. That behavior should be clear from the LAPACK documentation.

Comment (nalimilan) [inline base/linalg/lapack.jl]: Can't we always return a view, even if it covers the whole array when `n == size(A, 2)`? (I mean, after indexing is changed to return views.)

Comment (andreasnoack) [inline base/linalg/lapack.jl]: Yes. That's another type stable option.

**Comment by tkelman** [comment]
If I (or anyone else) backports this to release-0.4 at any point post-0.4.0, remember to revert 71dde8b6583489d4cdfe65689440f6f6bcf8b612 to put the docs back. Can't bootstrap if the docs are there but the methods aren't.

**Comment by ViralBShah** [comment]
Does anyone here want these LAPACK functions in 0.4? If so, the next RC may be the only realistic way to get them in. Since they are not being otherwise used by anything else in Base, unless someone really wants them, these can stay just in 0.5 and later.

**Pull request merged**